# Homework 10

In [1]:
import pandas as pd
import numpy as np
from scipy import linalg
import cvxpy as cp
import cvxopt

cp.installed_solvers()

['CVXOPT', 'ECOS', 'GLPK', 'GLPK_MI', 'OSQP', 'SCS']

### Question 1

#### Restricted Master Problem

$min \; x_1+x_2+x_3 \\
s.t. \;\; \begin{bmatrix}10\\0\\0\\\end{bmatrix}x_1+\begin{bmatrix}0\\7\\0\\\end{bmatrix}x_2+\begin{bmatrix}0\\0\\5\\\end{bmatrix}x_3 = \begin{bmatrix}15\\30\\20\\\end{bmatrix} \\
\;\;\;\;\;\;\;x_1,x_2,x_3\ge0$

**Optimal Solution**  
$\hat{x} = [\frac{3}{2},\frac{30}{7}, 4]$  

**Optimal Basis**  
$B = [A_1, A_2, A_3] = \begin{bmatrix}10&0&0\\0&7&0\\0&0&5\\\end{bmatrix},\;\;$ 
$c_B = \begin{bmatrix}1\\1\\1\\\end{bmatrix}$

$B^{-1} = \begin{bmatrix}\frac{1}{10}&0&0\\0&\frac{1}{7}&0\\0&0&\frac{1}{5}\\\end{bmatrix}\;\;$ 


$\hat{y} = c^T_BB^{-1} = \begin{bmatrix}\frac{1}{10}\\\frac{1}{7}\\\frac{1}{5}\\\end{bmatrix}$

### Question 2

In [2]:
# Solving using CVX
x1 = cp.Variable(name="x1")
x2 = cp.Variable(name="x2")
x3 = cp.Variable(name="x3")
objective = cp.Minimize(x1+x2+x3)

constraints = [10*x1 == 15,
               7*x2 == 30,
               5*x3 == 20,
               x1 >= 0,
               x2 >= 0,
               x3 >= 0
              ]

problem = cp.Problem(objective, constraints)

In [3]:
print("The optimal value is: " + str(problem.solve()))
print("The optimal solution is: [" + f'{x1.value:.2f}' + ", " + f'{x2.value:.2f}' + ", " + f'{x3.value:.2f}' + "]" )
print("The optimal dual solution is: [" + f'{-constraints[0].dual_value:.3f}' + ", " + f'{-constraints[1].dual_value:.3f}' + ", " + f'{-constraints[2].dual_value:.3f}' + "]" )

The optimal value is: 9.785714285714281
The optimal solution is: [1.50, 4.29, 4.00]
The optimal dual solution is: [0.100, 0.143, 0.200]


**Optimal Basis (same as above)**  
$B = [A_1, A_2, A_3] = \begin{bmatrix}10&0&0\\0&7&0\\0&0&5\\\end{bmatrix},\;\;$ 
$c_B = \begin{bmatrix}1\\1\\1\\\end{bmatrix}$

$B^{-1} = \begin{bmatrix}\frac{1}{10}&0&0\\0&\frac{1}{7}&0\\0&0&\frac{1}{5}\\\end{bmatrix}\;\;$ 


### Question 3

**Pricing Problem**

$\hat{Z} = max \;\frac{1}{10}a_1+\frac{1}{7}a_2+\frac{1}{5}a_3 \\
s.t. \;\; 7a_1+11a_2+16a_3 \le 80 \\
\;\;\;\;\;\;\;a_1, a_2, a_3 \ge 0, \; integers$

### Question 4

In [4]:
a1 = cp.Variable(name="a1", integer = True)
a2 = cp.Variable(name="a2", integer = True)
a3 = cp.Variable(name="a3", integer = True)
objective = cp.Maximize((1/10)*a1+(1/7)*a2+(1/5)*a3)

constraints = [7*a1+11*a2+16*a3 <= 80,
               a1 >= 0,
               a2 >= 0,
               a3 >= 0
              ]

problem = cp.Problem(objective, constraints)

In [5]:
print("The optimal value is: " + str(problem.solve()))
print("The optimal solution is: [" + str(a1.value) + ", " +str(a2.value) + ", " + str(a3.value) + "]" )

The optimal value is: 1.1
The optimal solution is: [11.0, 0.0, 0.0]


Since $\hat{Z}$ = 1.1, the minimum reduced cost is: $1-\hat{Z} = 1 - 1.1 = -0.1 \lt 0$.  
This means that the current solution $\hat{x} = [1, \frac{30}{7}, 4] $ is **not optimal** and basis column = [11,0,0] is added to the RMP. We don't have an assigned index, so I'll designate $A_4$ = [11,0,0].

### Question 5

#### Iteration 2: Restricted Master Problem

$min \; x_1+x_2+x_3+x_4 \\
s.t. \;\; \begin{bmatrix}10\\0\\0\\\end{bmatrix}x_1+\begin{bmatrix}0\\7\\0\\\end{bmatrix}x_2+\begin{bmatrix}0\\0\\5\\\end{bmatrix}x_3 +\begin{bmatrix}11\\0\\0\\\end{bmatrix}x_4= \begin{bmatrix}15\\30\\20\\\end{bmatrix} \\
\;\;\;\;\;\;\;x_1,x_2,x_3, x_4\ge0$

**Optimal Solution**  
$\hat{x} = [0, \frac{30}{7}, 4, \frac{15}{11}]$  

In [6]:
# Solving using CVX
x1 = cp.Variable(name="x1")
x2 = cp.Variable(name="x2")
x3 = cp.Variable(name="x3")
x4 = cp.Variable(name="x4")
objective = cp.Minimize(x1+x2+x3+x4)

constraints = [10*x1 + 11*x4 == 15,
               7*x2 == 30,
               5*x3 == 20,
               x1 >= 0,
               x2 >= 0,
               x3 >= 0,
               x4 >= 0
              ]

problem = cp.Problem(objective, constraints)

In [7]:
print("The optimal value is: " + str(problem.solve()))
print("The optimal solution is: [" + f'{x1.value:.2f}' + ", " + f'{x2.value:.2f}' + ", " + f'{x3.value:.2f}' +", " + f'{x4.value:.2f}' "]" )

The optimal value is: 9.649350650920342
The optimal solution is: [0.00, 4.29, 4.00, 1.36]


**Optimal Basis**  
$B = [A_2,A_3,A_4] = \begin{bmatrix}0&0&11\\7&0&0\\0&5&0\\\end{bmatrix},\;\;$ 
$c_B = \begin{bmatrix}1\\1\\1\\\end{bmatrix}$

$B^{-1} = \begin{bmatrix}0&\frac{1}{7}&0\\0&0&\frac{1}{5}\\\frac{1}{11}&0&0\\\end{bmatrix}\;\;$ 


$\hat{y} = c^T_BB^{-1} = \begin{bmatrix}\frac{1}{11}\\\frac{1}{7}\\\frac{1}{5}\\\end{bmatrix}$

In [8]:
B = np.array([[0,0,11], [7,0,0], [0,5,0]])
c = np.array([[1],[1], [1]])
B_inv = linalg.inv(B)
B_inv

array([[-0.        ,  0.14285714,  0.        ],
       [-0.        ,  0.        ,  0.2       ],
       [ 0.09090909,  0.        ,  0.        ]])

In [9]:
c.T@B_inv

array([[0.09090909, 0.14285714, 0.2       ]])

**Pricing Problem**

$\hat{Z} = max \;\frac{1}{11}a_1+\frac{1}{7}a_2+\frac{1}{5}a_3 \\
s.t. \;\; 7a_1+11a_2+16a_3 \le 80 \\
\;\;\;\;\;\;\;a_1, a_2, a_3 \ge 0, \; integers$

In [10]:
a1 = cp.Variable(name="a1", integer = True)
a2 = cp.Variable(name="a2", integer = True)
a3 = cp.Variable(name="a3", integer = True)
objective = cp.Maximize((1/11)*a1+(1/7)*a2+(1/5)*a3)

constraints = [7*a1+11*a2+16*a3 <= 80,
               a1 >= 0,
               a2 >= 0,
               a3 >= 0
              ]

problem = cp.Problem(objective, constraints)

print("The optimal value is: " + str(problem.solve()))
print("The optimal solution is: [" + str(a1.value) + ", " +str(a2.value) + ", " + str(a3.value) + "]" )

The optimal value is: 1.0389610389610389
The optimal solution is: [2.0, 6.0, 0.0]


Since $\hat{Z}$ = 1.04, the minimum reduced cost is: $1-\hat{Z} = 1 - 1.04 = -0.04 \lt 0$.  
This means that the current solution $\hat{x} = [0, \frac{30}{7}, 4, \frac{15}{11}] $ is **not optimal** and basis column = [2,6,0] is added to the RMP. We don't have an assigned index, so I'll designate $A_5$ = [2,6,0].

#### Iteration 3: Restricted Master Problem

$min \; x_1+x_2+x_3+x_4+x_5 \\
s.t. \;\; \begin{bmatrix}10\\0\\0\\\end{bmatrix}x_1+\begin{bmatrix}0\\7\\0\\\end{bmatrix}x_2+\begin{bmatrix}0\\0\\5\\\end{bmatrix}x_3 +\begin{bmatrix}11\\0\\0\\\end{bmatrix}x_4+\begin{bmatrix}2\\6\\0\\\end{bmatrix}x_5= \begin{bmatrix}15\\30\\20\\\end{bmatrix} \\
\;\;\;\;\;\;\;x_1,x_2,x_3, x_4,x_5\ge0$

**Optimal Solution**  
$\hat{x} = [0,  0,  4, \frac{5}{11}, 5] $  

In [11]:
# Solving using CVX
x1 = cp.Variable(name="x1")
x2 = cp.Variable(name="x2")
x3 = cp.Variable(name="x3")
x4 = cp.Variable(name="x4")
x5 = cp.Variable(name="x5")
objective = cp.Minimize(x1+x2+x3+x4+x5)

constraints = [10*x1 + 11*x4 + 2*x5 == 15,
               7*x2 + 6*x5 == 30,
               5*x3 == 20,
               x1 >= 0,
               x2 >= 0,
               x3 >= 0,
               x4 >= 0,
               x5 >= 0
              ]

problem = cp.Problem(objective, constraints)

In [12]:
print("The optimal value is: " + str(problem.solve()))
print("The optimal solution is: [" + f'{x1.value:.2f}' + ", " + f'{x2.value:.2f}' + ", " + f'{x3.value:.2f}' +", " + f'{x4.value:.2f}'+", " + f'{x5.value:.2f}' "]" )

The optimal value is: 9.454545477745505
The optimal solution is: [0.00, 0.00, 4.00, 0.45, 5.00]


**Optimal Basis**  
$B = [A_3, A_4,A_5] = \begin{bmatrix}0&11&2\\0&0&6\\5&0&0\\\end{bmatrix},\;\;$ 
$c_B = \begin{bmatrix}1\\1\\1\\\end{bmatrix}$

$B^{-1} = \begin{bmatrix}0&0&\frac{1}{5}\\\frac{1}{11}&-\frac{1}{33}&0\\0&\frac{1}{6}&0\\\end{bmatrix}\;\;$ 


$\hat{y} = c^T_BB^{-1} = \begin{bmatrix}\frac{1}{11}\\\frac{3}{22}\\\frac{1}{5}\\\end{bmatrix}$

In [13]:
B = np.array([[0,11,2], [0,0,6], [5,0,0]])
B_inv = linalg.inv(B)
B_inv

array([[ 0.        , -0.        ,  0.2       ],
       [ 0.09090909, -0.03030303,  0.        ],
       [ 0.        ,  0.16666667,  0.        ]])

In [14]:
c.T@B_inv

array([[0.09090909, 0.13636364, 0.2       ]])

**Pricing Problem**

$\hat{Z} = max \;\frac{1}{11}a_4+\frac{3}{22}a_2+\frac{1}{5}a_3 \\
s.t. \;\; 7a_1+11a_2+16a_3 \le 80 \\
\;\;\;\;\;\;\;a_1, a_2, a_3 \ge 0, \; integers$

In [15]:
a1 = cp.Variable(name="a1", integer = True)
a2 = cp.Variable(name="a2", integer = True)
a3 = cp.Variable(name="a3", integer = True)
objective = cp.Maximize((1/11)*a1+(3/22)*a2+(1/5)*a3)

constraints = [7*a1+11*a2+16*a3 <= 80,
               a1 >= 0,
               a2 >= 0,
               a3 >= 0
              ]

problem = cp.Problem(objective, constraints)

print("The optimal value is: " + str(problem.solve()))
print("The optimal solution is: [" + str(a1.value) + ", " +str(a2.value) + ", " + str(a3.value) + "]" )

The optimal value is: 1.0181818181818183
The optimal solution is: [9.0, 0.0, 1.0]


Since $\hat{Z}$ = 1.02, the minimum reduced cost is: $1-\hat{Z} = 1 - 1.02 = -0.02 \lt 0$.  
This means that the current solution $\hat{x} = [0, 0, 4, \frac{5}{11},5] $ is **not optimal** and basis column = [9,0,1] is added to the RMP. We don't have an assigned index, so I'll designate $A_6$ = [9,0,1].

#### Iteration 4: Restricted Master Problem

$min \; x_1+x_2+x_3+x_4+x_5+x_6 \\
s.t. \;\; \begin{bmatrix}10\\0\\0\\\end{bmatrix}x_1+\begin{bmatrix}0\\7\\0\\\end{bmatrix}x_2+\begin{bmatrix}0\\0\\5\\\end{bmatrix}x_3 +\begin{bmatrix}11\\0\\0\\\end{bmatrix}x_4+\begin{bmatrix}2\\6\\0\\\end{bmatrix}x_5+\begin{bmatrix}9\\0\\1\\\end{bmatrix}x_6= \begin{bmatrix}15\\30\\20\\\end{bmatrix} \\
\;\;\;\;\;\;\;x_1,x_2,x_3, x_4,x_5, x_6\ge0$

**Optimal Solution**  
$\hat{x} = [0, 0, \frac{35}{9}, 0, 5,\frac{5}{9}] $  

In [16]:
# Solving using CVX
x1 = cp.Variable(name="x1")
x2 = cp.Variable(name="x2")
x3 = cp.Variable(name="x3")
x4 = cp.Variable(name="x4")
x5 = cp.Variable(name="x5")
x6 = cp.Variable(name="x6")
objective = cp.Minimize(x1+x2+x3+x4+x5+x6)

constraints = [10*x1 + 11*x4 + 2*x5 + 9*x6 == 15,
               7*x2 + 6*x5 == 30,
               5*x3 + x6 == 20,
               x1 >= 0,
               x2 >= 0,
               x3 >= 0,
               x4 >= 0,
               x5 >= 0,
               x6 >= 0
              ]

problem = cp.Problem(objective, constraints)

In [17]:
print("The optimal value is: " + str(problem.solve()))
print("The optimal solution is: [" + f'{x1.value:.2f}' + ", " + f'{x2.value:.2f}' + ", " + f'{x3.value:.2f}' +", " + f'{x4.value:.2f}'+", " + f'{x5.value:.2f}'+", " + f'{x6.value:.2f}' "]" )

The optimal value is: 9.444444446310545
The optimal solution is: [0.00, 0.00, 3.89, 0.00, 5.00, 0.56]


**Optimal Basis**  
$B = [A_3, A_5,A_6] = \begin{bmatrix}0&2&9\\0&6&0\\5&0&1\\\end{bmatrix},\;\;$ 
$c_B = \begin{bmatrix}1\\1\\1\\\end{bmatrix}$

$B^{-1} = \begin{bmatrix}-\frac{1}{45}&\frac{1}{135}&\frac{1}{5}\\0&\frac{1}{6}&0\\\frac{1}{9}&-\frac{1}{27}&0\\\end{bmatrix}\;\;$ 


$\hat{y} = c^T_BB^{-1} = \begin{bmatrix}\frac{4}{45}\\\frac{37}{270}\\\frac{1}{5}\\\end{bmatrix}$

In [18]:
B = np.array([[0,2,9], [0,6,0], [5,0,1]])
B_inv = linalg.inv(B)
B_inv

array([[-0.02222222,  0.00740741,  0.2       ],
       [-0.        ,  0.16666667,  0.        ],
       [ 0.11111111, -0.03703704,  0.        ]])

In [19]:
c.T@B_inv

array([[0.08888889, 0.13703704, 0.2       ]])

**Pricing Problem**

$\hat{Z} = max \;\frac{4}{45}a_1+\frac{37}{270}a_2+\frac{1}{5}a_3 \\
s.t. \;\; 7a_1+11a_2+16a_3 \le 80 \\
\;\;\;\;\;\;\;a_1, a_2, a_3 \ge 0, \; integers$

In [20]:
a1 = cp.Variable(name="a1", integer = True)
a2 = cp.Variable(name="a2", integer = True)
a3 = cp.Variable(name="a3", integer = True)
objective = cp.Maximize((4/45)*a1+(37/270)*a2+(1/5)*a3)

constraints = [7*a1+11*a2+16*a3 <= 80,
               a1 >= 0,
               a2 >= 0,
               a3 >= 0
              ]

problem = cp.Problem(objective, constraints)

print("The optimal value is: " + str(problem.solve()))
print("The optimal solution is: [" + str(a1.value) + ", " +str(a2.value) + ", " + str(a3.value) + "]" )

The optimal value is: 1.0074074074074075
The optimal solution is: [6.0, 2.0, 1.0]


Since $\hat{Z}$ = 1.01, the minimum reduced cost is: $1-\hat{Z} = 1 - 1.01 = -0.01 \lt 0$.  
This means that the current solution $\hat{x} = [0, 0, \frac{35}{9}, 0,5,\frac{5}{9}] $ is **not optimal** and basis column = [6,2,1] is added to the RMP. We don't have an assigned index, so I'll designate $A_7$ = [6,2,1].

#### Iteration 5: Restricted Master Problem

$min \; x_1+x_2+x_3+x_4+x_5+x_6+x_7 \\
s.t. \;\; \begin{bmatrix}10\\0\\0\\\end{bmatrix}x_1+\begin{bmatrix}0\\7\\0\\\end{bmatrix}x_2+\begin{bmatrix}0\\0\\5\\\end{bmatrix}x_3 +\begin{bmatrix}11\\0\\0\\\end{bmatrix}x_4+\begin{bmatrix}2\\6\\0\\\end{bmatrix}x_5+\begin{bmatrix}9\\0\\1\\\end{bmatrix}x_6+\begin{bmatrix}6\\2\\1\\\end{bmatrix}x_7= \begin{bmatrix}15\\30\\20\\\end{bmatrix} \\
\;\;\;\;\;\;\;x_1,x_2,x_3, x_4,x_5, x_6,x_7\ge0$

**Optimal Solution**  
$\hat{x} = [0,0,\frac{61}{16},0,\frac{75}{16}, 0, \frac{15}{16} ] $  

In [21]:
# Solving using CVX
x1 = cp.Variable(name="x1")
x2 = cp.Variable(name="x2")
x3 = cp.Variable(name="x3")
x4 = cp.Variable(name="x4")
x5 = cp.Variable(name="x5")
x6 = cp.Variable(name="x6")
x7 = cp.Variable(name="x7")
objective = cp.Minimize(x1+x2+x3+x4+x5+x6+x7)

constraints = [10*x1 + 11*x4 + 2*x5 + 9*x6 + 6*x7 == 15,
               7*x2 + 6*x5 +2*x7 == 30,
               5*x3 + x6 + x7 == 20,
               x1 >= 0,
               x2 >= 0,
               x3 >= 0,
               x4 >= 0,
               x5 >= 0,
               x6 >= 0,
               x7 >= 0
              ]

problem = cp.Problem(objective, constraints)

In [22]:
print("The optimal value is: " + str(problem.solve()))
print("The optimal solution is: [" + f'{x1.value:.2f}' + ", " + f'{x2.value:.2f}' + ", " + f'{x3.value:.2f}' +", " + f'{x4.value:.2f}'+", " + f'{x5.value:.2f}'+", " + f'{x6.value:.2f}'+", " + f'{x7.value:.2f}' "]" )

The optimal value is: 9.437500001422672
The optimal solution is: [0.00, 0.00, 3.81, 0.00, 4.69, 0.00, 0.94]


**Optimal Basis**  
$B = [A_3, A_5,A_7] = \begin{bmatrix}0&2&6\\0&6&2\\5&0&1\\\end{bmatrix},\;\;$ 
$c_B = \begin{bmatrix}1\\1\\1\\\end{bmatrix}$

$B^{-1} = \begin{bmatrix} -\frac{3}{80}&\frac{1}{80}&\frac{1}{5} \\ 
                            -\frac{1}{16}&\frac{3}{16}&0 \\ 
                            \frac{3}{16}&-\frac{1}{16}&0 \\
              \end{bmatrix}\;\;$ 


$\hat{y} = c^T_BB^{-1} = \begin{bmatrix}\frac{7}{80}\\\frac{11}{80}\\\frac{1}{5}\\\end{bmatrix}$

In [23]:
B = np.array([[0,2,6], [0,6,2], [5,0,1]])
B_inv = linalg.inv(B)
B_inv

array([[-0.0375,  0.0125,  0.2   ],
       [-0.0625,  0.1875,  0.    ],
       [ 0.1875, -0.0625,  0.    ]])

In [24]:
c.T@B_inv

array([[0.0875, 0.1375, 0.2   ]])

**Knapsack Problem**

$\hat{Z} = max \;\frac{7}{80}a_1+\frac{11}{80}a_2+\frac{1}{5}a_3 \\
s.t. \;\; 7a_1+11a_2+16a_3 \le 80 \\
\;\;\;\;\;\;\;a_1, a_2, a_3 \ge 0, \; integers$

In [25]:
a1 = cp.Variable(name="a1", integer = True)
a2 = cp.Variable(name="a2", integer = True)
a3 = cp.Variable(name="a3", integer = True)
objective = cp.Maximize((7/80)*a1+(11/80)*a2+(1/5)*a3)

constraints = [7*a1+11*a2+16*a3 <= 80,
               a1 >= 0,
               a2 >= 0,
               a3 >= 0
              ]

problem = cp.Problem(objective, constraints)

print("The optimal value is: " + str(problem.solve()))
print("The optimal solution is: [" + str(a1.value) + ", " +str(a2.value) + ", " + str(a3.value) + "]" )

The optimal value is: 1.0
The optimal solution is: [0.0, 0.0, 5.0]


Since $\hat{Z}$ = 1.0, the minimum reduced cost is: $1-\hat{Z} = 1 - 1.0 = 0 \lt 0$.  
This means that the current solution $\hat{x} = [0,0,\frac{61}{16},0,\frac{75}{16}, 0, \frac{15}{16} ] $  is **indeed optimal.**

### Question 6

The final optimal solution is $\hat{x} = [0,0,\frac{61}{16},0,\frac{75}{16}, 0, \frac{15}{16} ] $  
The optimal basis is $B = [A_3, A_5,A_7] = \begin{bmatrix}0&2&6\\0&6&2\\5&0&1\\\end{bmatrix}\;\;$   
The optimal objective value is 9.44.